### PR TITLE
Test Automation for Calculated Columns

### PR DESCRIPTION
**Pull request by labkey-danield**
#### Rationale
Updated some helpers for calculated columns tests in LIMS module.
Added a test in the LKS domain designer.

#### Related Pull Requests
* https://github.com/LabKey/limsModules/pull/661

#### Changes
* Updated DomainFieldRow:
    * Remove redundant code that was setting format.
    * Added method to set and get the calculated column expression.
    * Added a getter for the warning message.
    * Added a getter to see if 'Required Field' checkbox is present.
* Added Calculation column type.
* Added DomainDesignerTest.testCalculatedColumn
    * Validates that setting attributes before changing field to a Calculated type.

### CODE REVIEW

File: src/org/labkey/test/components/domain/DomainFieldRow.java [hunk 1]
```diff
@@ -916,6 +975,40 @@ public DomainFieldRow deleteTextChoiceValue(String value)
...
+    }
+
+    public String getValueExpressionStatusMessage()
+    {
+        // Need to remove focus from the expression field to have the updated status message show up.
+        getWrapper().fireEvent(elementCache().expressionInput, WebDriverWrapper.SeleniumEvent.blur);
+        WebDriverWrapper.sleep(500);
+
+        String statusMsg = "";
+        if(Boolean.TRUE.equals(waitFor(elementCache().expressionStatusMsg::isDisplayed, 1_000)))
```
Comment (labkey-tchad) [inline src/org/labkey/test/components/domain/DomainFieldRow.java]: No need to do `TRUE.equals` here, `WebElement.isDisplayed()` won't ever return `null`.
```suggestion
        if(waitFor(elementCache().expressionStatusMsg::isDisplayed, 1_000))
```

File: src/org/labkey/test/params/FieldDefinition.java [hunk 2]
```diff
@@ -509,7 +509,7 @@ public boolean isMeasureByDefault()
         ColumnType Barcode = new ColumnTypeImpl("Unique ID", "string", "http://www.labkey.org/types#storageUniqueId", null);
         ColumnType TextChoice = new ColumnTypeImpl("Text Choice", "string", "http://www.labkey.org/types#textChoice", null);
         ColumnType SMILES = new ColumnTypeImpl("SMILES", "string", "http://www.labkey.org/exp/xml#smiles", null);
-
+        ColumnType Calculation = new ColumnTypeImpl("Calculation", "http://www.labkey.org/exp/xml#calculated");
```
Comment (labkey-tchad) [inline src/org/labkey/test/params/FieldDefinition.java]: "`http://www.labkey.org/exp/xml#calculated`" is the `conceptURI` for this type, not the `rangeURI`. I think we want something like this:
```suggestion
        ColumnType Calculation = new ColumnTypeImpl("Calculation", null, "http://www.labkey.org/exp/xml#calculated", null);
```
Though it doesn't really matter unless we add API support for calculated columns

**Review by labkey-tchad — APPROVED** [review]
Would it be worth adding API support for calculated columns on the test side? It would allow us to do more thorough testing of various calculations without having to go through the designer.

**Review by cnathe — APPROVED** [review]

File: src/org/labkey/test/components/domain/DomainFieldRow.java (hunk 1 above)
Comment (labkey-danield) [reply src/org/labkey/test/components/domain/DomainFieldRow.java]: Changed.

File: src/org/labkey/test/params/FieldDefinition.java (hunk 2 above)
Comment (labkey-danield) [reply src/org/labkey/test/params/FieldDefinition.java]: Changed.
The plan is to add API support for calculated columns. I have to open a bug and assign it to Cory.

**Pull request merged**